test/lib/faktory/connection_test.exs

```diff
@@ -31,7 +31,7 @@ defmodule Faktory.ConnectionTest do
       wid: "123abc",
       password: nil
     })
-    :timer.sleep(10)
+    :timer.sleep(10) # Ugh, I hate this.
```

Comment:
How would you feel about passing a PID (`self()` in this case) into `Connection.start_link` and then having `Connection` send a message like `:ready` to that PID once the handshake completes? Then here and elsewhere, we can do a `receive` to block until things are ready to roll - or fail if it times out.

Reply:
I've worked in the Ruby/Rails world for so long now that I can't get a feel for if it's good or bad to write code that explicitly facilitates testing. And by explicit, I mean `if env == :test` type stuff. In Ruby, the mocking/stubbing is so powerful that your app code is usually blissfully unaware of any testing.

So taking this idea to the logical conclusion, almost all the functions in `Protocol` would have stuff like:
```
if_test do
  send conn.test_pid, :fetch_done
end
```
at the end of them.

Is this normal or considered good practice? I'm learning towards yes? But don't know for sure.

Reply:
Actually... this code is more about communicating the steps in an async process. Not exactly mocking.

Still littering app code with`if_test, do: ...` feels weird to me. But I think maybe it shouldn't.

Reply:
It's a good question. In general, I don't think it's good hygiene to sprinkle testing code in app code unless you're in a tight spot and have no other options.

I was originally thinking that the callback message could be useful in app code as well, since you don't want to risk using the connection before the handshake has finished. Related to that, I was just reading some of the code in the Connection lib and came across [this](https://github.com/fishcakez/connection/blob/master/lib/connection.ex#L103-L104):

> This return value is useful when the process connects inside `init/1` so that
  `start_link/3` blocks until a connection is established.

I like that. The TCP connection startup should be very fast, and the handshake will only require ~3 roundtrips, so doing it inside of `init` shouldn't cause a timeout. That would eliminate the need for these hacky workarounds and should make the app code more predictable too.

Thoughts?

Reply:
The reason why the handshake is in the `connect/2` callback is so that if the socket gets disconnected after having a successful handshake, that it can just go straight back to that function to re-handshake after it reestablishes connection.

Now that I say that out loud (in my head) and look at the code... there is no reason it can't do the handshake in `init/1` _and_ in the reconnect codepath. Alright, I'm with ya.

And now looking at `Protocol`, everything is synchronous... I was wrong about having to sprinkle it with `if_test do: send(test_pid, :fetch_done` type stuff.